File: LeetCode/1477. Find Two Non-overlapping Sub-arrays Each With Target Sum/solution.py
# class Solution:
    # def minSumOfLengths(self, arr: List[int], target: int) -> int:
    #     res = float('inf')
    #     ends = [-1]
    #     mp = {-1:float('inf')}
    #     left, right, sums, n = 0, 0, 0, len(arr)
    #     cur = 0
    #     for right in range(n):
    #         sums += arr[right]
    #         while left < n and sums > target:
    #             sums -= arr[left]
    #             left += 1
    #             while cur+1 < len(ends) and left > ends[cur+1]:
    #                 cur += 1
    #         if sums == target:
    #             res = min(res, mp[ends[cur]]+right-left+1)
    #             if right-left+1 < mp[ends[-1]]:
    #                 ends.append(right)
    #                 mp[right] = right-left+1
    #             # print(ends,left,right,cur)
    #     return -1 if res == float('inf') else res
    
class Solution:
    def minSumOfLengths(self, arr: List[int], target: int) -> int:
        prefix = {0: -1}
        best_till = [math.inf] * len(arr)
        ans = best = math.inf
        for i, curr in enumerate(itertools.accumulate(arr)):
            if curr - target in prefix:
                end = prefix[curr - target]
                if end > -1:
                    ans = min(ans, i - end + best_till[end])
                best = min(best, i - end)
            best_till[i] = best
            prefix[curr] = i
        return -1 if ans == math.inf else ans

class Solution:
    def minSumOfLengths(self, a: List[int], targ: int) -> int:
        
        n = len(a)
        d1 = defaultdict(lambda: inf)
        ret2 = inf
        tot = 0
        j = 0
        
        for i,x in enumerate(a):
            
            tot += x
            
            while tot > targ:
                tot -= a[j]
                j += 1
            
            sz = i-j+1
            
            if tot == targ:
                d1[i] = sz
                ret2 = min(ret2, d1[j-1] + sz)
            
            d1[i] = min(d1[i], d1[i-1])
        
        return -1 if ret2 == inf else ret2

File: LeetCode/1477. Find Two Non-overlapping Sub-arrays Each With Target Sum/solution.cpp
// const int inf = 0x3f3f3f3f;
// class Solution {
// public:
//     int minSumOfLengths(vector<int>& arr, int target) {
//      int n = arr.size();
//      vector<int> left(n,inf), right(n,inf);
//      unordered_map<int,int> mp;
//      int sums = 0;
//      mp[0] = -1;
//      for (int i = 0; i < n; i++){
//          sums += arr[i];
//          if (mp.count(sums-target))
//              left[i] = i-mp[sums-target];
//          if (i > 0)
//              left[i] = min(left[i], left[i-1]);
//          mp[sums] = i;
//      }
//      sums = 0;
//      mp.clear();
//      mp[0] = n;
//      for (int i = n-1; i >= 0; i--){
//          sums += arr[i];
//          if (mp.count(sums-target))
//              right[i] = mp[sums-target]-i;
//          if (i < n-1)
//              right[i] = min(right[i], right[i+1]);
//          mp[sums] = i;
//      }
//      int res = inf;
//      for (int i = 1; i < n; i++){
//          res = min(res, left[i-1]+right[i]);
//      }
//      return res == inf ? -1 : res;

//     }
// };


// const int inf = 0x3f3f3f3f;
// class Solution {
// public:
//     int minSumOfLengths(vector<int>& arr, int target) {
//      int n = arr.size();
//      int res = inf;
//      vector<int> left(n,inf);
//      unordered_map<int,int> mp;
//      int sums = 0;
//      mp[0] = -1;
//      for (int i = 0; i < n; i++){
//          sums += arr[i];
//          if (mp.count(sums-target)){
//              if (mp[sums-target] >= 0)
//                  res = min(res, left[mp[sums-target]]+i-mp[sums-target]);
//              left[i] = i-mp[sums-target];
//          }
//          if (i > 0)
//              left[i] = min(left[i], left[i-1]);
//          mp[sums] = i;
//      }
//      return res == inf ? -1 : res;

//     }
// };



// const int inf = 0x3f3f3f3f;
class Solution {
public:
    int minSumOfLengths(vector<int>& arr, int target) {
        int n = arr.size();
        int res = inf;
        vector<int> dp(n,inf);
        int sums = 0, left = 0, right = 0;
        for (; right < n; right++){
            sums += arr[right];
            while (sums > target)
                sums -= arr[left++];
            if (sums == target){
                dp[right] = right-left+1;
                if (left > 0)
                    res = min(res, right-left+1+dp[left-1]);
            }
            if (right > 0)
                dp[right] = min(dp[right], dp[right-1]);
        }
        return res == inf ? -1 : res;

    }
};

// const int inf = 0x3f3f3f3f;
// const int maxn = 1e5+5;

// int dp[maxn];

// class Solution {
// public:
//     int minSumOfLengths(vector<int>& arr, int target) {
//      int n = arr.size();
//      int res = inf;
//      int sums = 0, left = 0, right = 0;
//      memset(dp, 0x3f, n*sizeof(int));
//      for (; right < n; right++){
//          sums += arr[right];
//          while (sums > target)
//              sums -= arr[left++];
//          if (sums == target){
//              dp[right] = right-left+1;
//              if (left > 0)
//                  res = min(res, right-left+1+dp[left-1]);
//          }
//          if (right > 0)
//              dp[right] = min(dp[right], dp[right-1]);
//      }
//      return res == inf ? -1 : res;

//     }
// };